src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Tree/ProjectImports/ProjectImportsSubTreeProvider.cs

```diff
@@ -0,0 +1,330 @@
+                }
+
+                ToggleTree();
+
+                return;
+
+                void ToggleTree()
+                {
+                    // Queue up an operation that will toggle the import tree
+                    _unconfiguredProjectTasksService.LoadedProjectAsync(
+                        async () =>
+                        {
+                            await TaskScheduler.Default.SwitchTo(alwaysYield: true);
+
+                            _unconfiguredProjectTasksService.UnloadCancellationToken.ThrowIfCancellationRequested();
+
+                            lock (SyncObject)
+                            {
+                                Verify.NotDisposed(this);
+
+                                // Use the presence or absence of a subscription to indicate which operation we are performing here.
+                                // This avoids a race condition between the (locked) changing of _showAllFiles and the (locked) changing
+                                // of _subscriptions. Even if there is a race, the right number of toggles will occur and the end result
+                                // will be correct.
+                                if (_subscriptions == null)
+                                {
+                                    SetUpTree();
+                                }
+                                else
+                                {
+                                    TearDownTree();
+                                }
+                            }
+                        });
+
+                    return;
+
+                    void SetUpTree()
+                    {
+                        Assumes.Null(_subscriptions);
+
+                        // Set a visible root
+                        SubmitTreeUpdateAsync(
+                            (currentTree, configuredProjectExports, token) =>
+                            {
+                                // Update (make visible) or create a new tree if no prior one exists
+                                IProjectTree tree = currentTree == null
+                                    ? NewTree(Resources.ImportsTreeNodeName, icon: s_rootIcon, flags: s_projectImportsTreeRootFlags)
+                                    : currentTree.Value.Tree.SetVisible(true);
+
+                                return Task.FromResult(new TreeUpdateResult(tree));
+                            });
+
+                        // Subscribe to data to populate the tree and keep it updated with changes
+                        IReceivableSourceBlock<IProjectVersionedValue<IProjectImportTreeSnapshot>> importTreeSource
+                            = _projectSubscriptionService.ImportTreeSource.SourceBlock;
+
+                        IReceivableSourceBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>> projectRuleSource
+                            = _projectSubscriptionService.ProjectRuleSource.SourceBlock;
+
+                        var intermediateBlock =
```

Comment:
What does this block actually do?

Reply:
You need an intermediate block to allow filtering to specific rule names via `LinkTo` below.